### ISSUE TITLE
vsce create-publisher returns 401

### ISSUE THREAD

**Issue by SamVerschueren** [Microsoft/vscode-vsce#11]
I'm trying to create a publisher with a `Personal Access Token` but it fails to create it.

> Error: Failed Request: Unauthorized(401) - https://app.market.visualstudio.com/_apis/gallery

**Comment by Bigous** [comment]
Hi @SamVerschueren , where did you change this setting?

**Comment by amamut** [comment]
Adding a link to explain what this means as I googled this issue and landed here. No real explanation was given as how to enable this setting.

https://code.visualstudio.com/docs/tools/vscecli#_common-questions

**Comment by joaomoreno** [comment]
Thanks, @amamut!

Ping @waderyan: we really need SEO on our docs.

**Comment by waderyan** [comment]
@pjmeyer see above. Could use your help on improving SEO holistically for the site.

**Comment by Sequoia** [comment]
I'm struggling with this now, can't figure out what I'm supposed to enter as publisher for `create-publisher` and the docs all link in a circle 😦

**Comment by joaomoreno** [comment]
Hi @Sequoia, sorry to read that! Could you get a hold of a [Personal Access Token](https://code.visualstudio.com/docs/tools/vscecli)? Once you do, you can call `vsce create-publisher publishername` to create it. Do you see any errors?

**Comment by Sequoia** [comment]
@joaomoreno Yes I have one, what i don't know is what to enter for `publishername`. Is it my "Team" name? An arbitrary string? I just keep getting 401 with no additional info.

**Comment by Sequoia** [comment]
Here's my "team" URL: https://34ks.visualstudio.com/

**Comment by npruehs** [comment]
Same here.

OS Version: Windows 10 Home 10.0.10538 

VS Team Services Access Token Details:

![vsteamservices](https://cloud.githubusercontent.com/assets/757321/17086598/cc988832-51f6-11e6-82a6-6e3257626fa4.png)

Output:

```
C:\Users\Nick\Desktop\pony>vsce create-publisher npruehs
Publisher human-friendly name: (npruehs)
Personal Access Token: ****************************************************

Error: Failed Request: Unauthorized(401) -
```

**Comment by joaomoreno** [comment]
@npruehs This issue is closed.

Are you sure you created your PAT for **all** accessible accounts?

> A: One easy mistake to make when creating the PAT (Personal Access Token) is to not select all accessible accounts in the Accounts field dropdown (instead selecting a specific account). You should also set the Authorized Scopes to All scopes for the publish to work.

**Comment by Sequoia** [comment]
@joaomoreno this bug should be migrated to the VSCode Extension *platform/server* repo; this is a UI bug in the server's response.

## Expected
`401 Unauthorized. If you are receiving this response on your own account, see faq.vscode.biz/#vsce-unauthorized`

User is given some instruction on how to remedy this 👉🏼*common*👈🏼 issue with the peculiarities of the MSFT/VS permissions system that are frequently confusing to new users.

## Actual
`401 unauthorized`

User is given no help, it appears they entered something wrong *at the command line* (whereas in fact the issue is with server account settings). You try to fix it, checking your credentials and retrying several times, to no avail. 😿

**Comment by Sequoia** [comment]
Or perhaps the help message should come from `vsce` upon 401? Either way, the user should be given *some* sort of guidance to help them handle this issue, which is a roadblock in particular for new users.

**Comment by joaomoreno** [comment]
Agreed. Pushed a commit which guides users towards fixing this specific problem.

**Comment by npruehs** [comment]
Worked like a charm. Thanks all!

**Comment by Sequoia** [comment]
👏🏼👏🏼👏🏼👍🏼 thanks!!

**Comment by joshjg** [comment]
I get 401 here even though I set up the token for all accessible accounts. At a dead end here...

**Comment by joaomoreno** [comment]
@joshjg Have you tried picking another publisher name?

**Comment by joshjg** [comment]
@joaomoreno Yes, to no avail.

**Comment by joaomoreno** [comment]
@modigrv @hamenon Some help here?

**Comment by hamenon** [comment]
@joshjg could you give us some more info like the account name where you created the token and the publisher id you are looking to create? Might help us investigate further. If possible can you write to us at vsmarketplace@microsoft.com we can have someone from the team debug the issue. 

As a workaround you could create the publisher manually at http://marketplace.visualstudio.com/manage and upload the packaged vsix directly to get unblocked

**Comment by ipedrazas** [comment]
So, I'm in the same situation, trying `vsce create-publisher` and getting a 401.

I've just published the extension using @hamenon workaround 👏 

Why don't you put this workaround in the docs?

**Comment by joaomoreno** [comment]
Added to the docs: https://github.com/Microsoft/vscode-docs/commit/5fad3ce7dcec9ed3ddbf4b9bd6f172174f499785

**Comment by ipedrazas** [comment]
I'm afraid the workaround only allows you to create a new extension but not to update one.

What happens if I delete my extension and publish a new version? how this affects users with the extension installed?

How can I debug `vsce`? I've tried both `vsce login` and `vsce create-publisher` and I've got the same error (401) in both cases.

**Comment by ipedrazas** [comment]
I've managed to update the extension using the very hidden option in the extension... If you point me to the GitHub repo of the marketplace I'd like to file an issue.

**Comment by andradei** [comment]
This is such a huge pain point to beginner extension devs. If it weren't for this issue, I'd be stuck.

**Comment by joaomoreno** [comment]
@hamenon There seems to be a lot of friction over here. Is there any way vsce can get the error message from the 401 through the use of `vso-node-api`?

**Comment by hamenon** [comment]
We did discover an issue with the PAT's that are issued by the VSTS Identity Service. This has now been fixed and the deployed on production. Note that only PAT's created in the last week or so are affected. 

1. Existing users with Old PAT's (over a week old) can continue using them
2. Existing users who created new  PAT's in the last week need to generate new PAT's
3. New users trying to create publishers can generate PAT now and get going.

**Comment by geddski** [comment]
So I just hit this trying to publish my first extension. Really have no idea what "which has access to **all accessible accounts**" means. What should I do from here?

**Comment by joshjg** [comment]
@geddski This is the option you're looking for:
<img width="420" alt="screen shot 2016-11-17 at 8 57 58 pm" src="https://cloud.githubusercontent.com/assets/21149010/20415795/aacee9ea-ad08-11e6-876c-63ea1c8f32ab.png">

**Comment by geddski** [comment]
Ah that worked. Thanks!

**Comment by AvraamMavridis** [comment]
I am trying the same but I dont have the all accessible acounts option...

<img width="511" alt="screen shot 2018-07-31 at 21 48 09" src="https://user-images.githubusercontent.com/283690/43483289-8297e66a-950b-11e8-97ca-c3949558caba.png">

**Comment by cuiyongjian** [comment]
i also have created token at azure devops, but i get:
```
Publisher human-friendly name: (niubiheader)
E-mail: cuiyongjian@outlook.com
Personal Access Token: ****************************************************

(node:65835) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
Error: Failed Request: Unauthorized(401) - https://marketplace.visualstudio.com/_apis/gallery
Be sure to use a Personal Access Token which has access to **all accessible accounts**.
See https://code.visualstudio.com/docs/tools/vscecli#_common-questions for more information.
```

**Comment by toreaurstad** [comment]
Well, I am using an account at work and I had to enable a PAT for "All accessible organizations", which of course is similar to the case "All accounts". But it should be noted as a possible scenario.

**Comment by guitartsword** [comment]
vsce --version
1.66.0

The message is still: No guidance to user 😢 
![image](https://user-images.githubusercontent.com/12071733/61753152-8ff7d300-ad6b-11e9-98cc-10b27a8bc256.png)

I installed it today via `npm i -g vsce`

**Comment by zhohon** [comment]
> The message is still not helping the user😢

+1

**Comment by zhohon** [comment]
> vsce --version
> 1.66.0
> 
> The message is still not helping the user😢
> ![image](https://user-images.githubusercontent.com/12071733/61753152-8ff7d300-ad6b-11e9-98cc-10b27a8bc256.png)
> 
> I installed it today via `npm i -g vsce`

+1

**Comment by creaink** [comment]
> vsce --version
> 1.66.0
> 
> The message is still not helping the user😢
> ![image](https://user-images.githubusercontent.com/12071733/61753152-8ff7d300-ad6b-11e9-98cc-10b27a8bc256.png)
> 
> I installed it today via `npm i -g vsce`

use `vsce login` and paste token is fine for me

**Comment by MaskRay** [comment]
Got the same unhelpful error message.
```
 ERROR  Failed request: (401)
```

On `https://dev.azure.com/$myorganization/_usersSettings/tokens`, changing the scope to "All accessible organizations" fixed the 401 problem.

**Comment by syyam** [comment]
Select this option from the dropdown.
![image](https://user-images.githubusercontent.com/20249719/69012128-2b085e80-0994-11ea-964f-ffed6b908c5b.png)

**Comment by erlangp** [comment]
Thank you very much @hamenon

**Comment by vsTianhao** [comment]
It was a terrible mistake, I don't understand the meaning of choice  “All accessible organizations” , The other option doesn't seem to make any sense.

It's not a lack of SEO docs, it's a bug

**Comment by ngarg-panw** [comment]
I have lost the existing token. The person who generated the token first time is no longer with the team. How can I associate a new token from a new account to existing publisher?
I tried creating a token from new account and it says:
ERROR Access Denied: abcxyzteam needs the following permission(s) on the resource /Pxxxxx to perform this action: View user permissions on a resource
Any pointers are helpful.

**Comment by teerasej** [comment]
> I have lost the existing token. The person who generated the token first time is no longer with the team. How can I associate a new token from a new account to existing publisher?
> I tried creating a token from new account and it says:
> ERROR Access Denied: abcxyzteam needs the following permission(s) on the resource /Pxxxxx to perform this action: View user permissions on a resource
> Any pointers are helpful.

Got the same thing. I tried to find that permission in the token list, but unlucky.

**Comment by sw-yx** [comment]
hey @maintainers pls put this in your error message or make your UI more intuitive.

**Comment by lukepighetti** [comment]
Please just give us a one click oauth system. I have no clue what my publisher name is. My token isn't working. I can't figure this out. The docs go in a loop. There is no example of what "publisher" is supposed to look like. I have never in my life seen an auth flow so complicated, even on dev tooling.

**Comment by andradei** [comment]
At least a video or blog with screenshots would be a good step towards making this process less of a pain. Right now it sucks big time.

**Comment by lukepighetti** [comment]
It's the next day and I've given up on publishing from the terminal. Guess I'm rocking the Marketplace update dropzone from here on out.

**Comment by william-takayama** [comment]
It worked for me just when I've allowed `full access` option
![image](https://user-images.githubusercontent.com/60292132/121973127-f8cb0880-cd52-11eb-902a-94a9d4d435a5.png)

**Comment by erlangparasu** [comment]
> I am trying the same but I dont have the all accessible acounts option...
> 
> <img alt="screen shot 2018-07-31 at 21 48 09" width="511" src="https://user-images.githubusercontent.com/283690/43483289-8297e66a-950b-11e8-97ca-c3949558caba.png">

**All accessible organizations**
thanks it works

**Comment by bennycode** [comment]
Here is how I solved the 401 error in order to publish VS Code extensions using a GitHub Actions workflow:

1. I set "Organiation" to "All accessible organizations" for my Personal Access Token (PAT)
1. I used a custom defined scope of type "Marketplace: Manage"
1. I set the expiry date to the future
1. I created a GitHub Repository secret with the name of `VSCE_PAT` (as it is the [default env variable](https://github.com/microsoft/vscode-vsce/blob/v2.15.0/src/main.ts#L170) used when publishing)
1. I created the following workflow commands:

```yml
- name: 'Install dependencies'
  run: |
    yarn global add @vscode/vsce
    yarn install
- name: 'Authenticate'
  run: |
    git config --global user.email "${{ secrets.MY_GITHUB_EMAIL }}"
    git config --global user.name "Benny Neugebauer"
- name: 'Publish VS Code extension'
  env:
    VSCE_PAT: ${{ secrets.VSCE_PAT }}
  run: |
    vsce publish patch
```

I added a "postversion" script to my "package.json" file to push the version update also to my GitHub repository:

```json
"scripts": {
  "postversion": "git push origin main && git push origin --tags"
}
```

Full version here: https://github.com/bennycode/sort-everything/blob/v1.3.3/.github/workflows/publish-extension.yml

Hope this helps! ✌️
